### ISSUE TITLE
Seamless tile loading

### ISSUE THREAD

**Issue by Nohus** [peterLaurence/MapCompose#29]
Currently tile loading is very visible for the user. When the map is scrolled to a new tile, it's always empty, has to load in, and them run the fade-in animation. I would expect to be able to scroll without ever seeing any loading (unless scrolling very fast), especially when tiles are locally-cached and don't need to load from the network.

The solution is to not only load visible tiles, but also have a margin of off-screen tiles loaded. This margin could be configurable (and default to 0, if you think the current behavior should stay the default). Also, when a tile is loaded off-screen, it should not run the fade-in animation.

I'd be happy to implement this myself, but there are significant changes happening on the `layers` branch, so there might be conflicts. What do you think?

**Comment by peterLaurence** [comment]
A simple approach would be to be to modify the `VisibileTiles` before it's passed to the `VisibleTilesResolver`. By applying a padding (passed as dp amount with an api) would virtually increase the visible area, causing tiles on the borders to be preloaded. The default should remain 0. I don't think it's necessary to disable fadein for non visible tiles. The android canvas already discards all drawing operations which aren't visible. 
However, work on `layers` should be finished first. It would be good feature you could add.

**Comment by Nohus** [comment]
> I don't think it's necessary to disable fadein for non visible tiles. The android canvas already discards all drawing operations which aren't visible.

The concern is not performance due to doing off-screen animations, but that if the off-screen tile loads, starts the animation, then immediately after the user scrolls that way, they will see the animation, even though it wasn't necessary.

**Comment by peterLaurence** [comment]
I'm not sure I follow. My previous library, MapView, had this feature. In practice, with a padding large as 2 or 3 tiles, you don't see any animation because their animation is already done. Animation isn't per screen, it's per tile.

**Comment by Nohus** [comment]
> In practice, with a padding large as 2 or 3 tiles, you don't see any animation because their animation is already done.

Sure, maybe I am overthinking this. But in theory with a small enough padding or high enough scrolling speed you would see it. Imagine an instant `TileStreamProvider` with a fast scrolling map. In that setup you could never see any loading, even with a padding of just 1 tile. But if the animations still play, you will see them as you scroll into the padding tiles.

**Comment by peterLaurence** [comment]
You might think that the animation starts when the tile is visible. It starts when it's added to the list of tiles to be drawn. In practice, a padding and even making the fadin more quick does the trick.

**Comment by peterLaurence** [comment]
Yeah trust me, the padding is enough. An even in case you're scrolling like a mad person, you barely see the end of a fadein on the borders...

**Comment by peterLaurence** [comment]
Somewhat related to this feature proposal, I'm curious to know: are you fetching your tiles from http request?
If so, how many workers have you set congured in your `MapState`?

**Comment by Nohus** [comment]
Sorry, I didn't get a notification for some reason. Yes I'm fetching tiles from a server, and have configured 16 workers. But not there wasn't much thought behind that value. I was going to test different values once my code is ready and I have layers implemented.

**Comment by peterLaurence** [comment]
16 workers might be good enough. However the optimal number depends on the average response time of your servers, the device, and even the tile size.

**Comment by peterLaurence** [comment]
Since everything was already in place to set a padding, I just went ahead and created a `setPadding` API.
You can test it in a demo by adding `setPadding(600)`.
It's dynamic. We can invoke `setPadding(600)` and later on `setPadding(0)` with the expected effect.
Let me know if that's good enough for you.

**Comment by Nohus** [comment]
Thank you, I will test it now. I also got an email from GitHub with a link to https://github.com/peterLaurence/MapCompose/discussions/30 but it seems gone?

**Comment by peterLaurence** [comment]
Yes, I removed it as I thought you were not interested in this project anymore. I can see the misunderstanding now. Basically, I told you that I created a `v2` branch in which I merged `master`. So `v2` has all your contributions and the layers feature.
In the `v2` branch, I started to update the readme and describe the layers feature. I should also add an acknowledgement section (mentioning you as contributor and tester).
The `2.0.0-beta01` was published yesterday on MavenCentral. It doesn't have my last modification for seamless tile loading though. 
That will be part of the next beta version.

**Comment by Nohus** [comment]
> I thought you were not interested in this project anymore

I was just waiting for the layers to be ready for testing. :smile: I will go on and give it a go now.

**Comment by peterLaurence** [comment]
No worries. I've seen so many people come and go, sometimes never answering all of a sudden.

**Comment by Nohus** [comment]
This works with the new `setPadding()`, thanks!
